Add `Breadcrumbs` component

ntsekouras opened the pull request:
Resolves: DES-99

This PR implements the `Breadcrumbs` component. It has an `isCompact` property to force the component to be rendered in a compact format. If not set to `true` the component switches to compact mode and back based on the available space.

My approach for detecting when to switch to compact and back is always rendering an `offscreen` and full width version of the breadcrumbs and compare its size with current rendered (and visible) instance.


## Notes

In this PR I'm also unlocking and exporting the `Menu` core component.

## Testing Instructions
1. in storybook `yarn components:storybook:start`


### Screenshot


https://github.com/user-attachments/assets/dd5e3f6f-c590-40d4-a6d3-0f54a17e3720




<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

*

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

ntsekouras committed:
Add `Breadcrumbs` component

jasmussen commented:
This looks good. We were discussing the precise behavior of breadcrumbs in a Figma file (s2MBbwtgu3uhMQE3zsoUoT-fi-0_1#1231866885), and the responsive behavior you've built here adds some valuable nuance. 

Take this example which uses a header component:

![Screenshot 2025-04-30 at 10 46 57](https://github.com/user-attachments/assets/38dd7a87-f8fe-4615-aea9-7d7cdf88e524)

Here, the "General" breadcrumb is duplicated as the page title, right below, which is unfortunate. One proposal there is to hide the breadcrumb, optionally, so you could have something like this:

![Screenshot 2025-04-30 at 10 46 44](https://github.com/user-attachments/assets/5cbab4d8-007f-42bc-b601-8bd90ae20b25)

This is a challenge of course, if the second to last breadcrumb elides to [...] on smaller screens. 

Besides input from @jameskoster, I'd love insights from @StevenDufresne in case he has time, as I think he was involved in breadcrumbs on the WordPress.org website, including eliding heuristics. From revisiting this, I don't think we actually do elide, but we do horizontal scrolling with a scrim on the right. [Example](https://developer.wordpress.org/advanced-administration/security/hardening/display-errors/).

jasmussen commented:
An additional thought from discussing in the Figma file. If we are going to elide like this, we should probably refine the behavior so that it never elides the parent breadcrumb, because the parent page (and the root page) is arguably the most common section you want to go back to.

That means in your crumbs example we'd have this:

> Home / ... / Gaming / Alienware X17 

That also means if there's only ever three breadcumbs (e.g. Home / Products / Electronics), we would _not_ elide.

If we were to adopt this approach, that would still allow a prop to hide the last breadcrumb in cases where the title of the page is shown right below.

youknowriad commented on packages/components/src/breadcrumbs/types.ts:
```diff
@@ -0,0 +1,36 @@
+	 * Note that when `false` the current page is only visually hidden.
+	 * @default true
+	 */
+	showCurrentPage?: boolean;
```
if the items are called "items", should this one be called `showActiveItem` or `showCurrentItem`?

youknowriad commented on packages/components/src/breadcrumbs/types.ts:
```diff
@@ -0,0 +1,36 @@
+	 * and back based on the available space.
+	 * @default false
+	 */
+	isCompact?: boolean;
```
I wonder if this should be `size?: 'default' | 'compact'` like we have in a few `@wordpress/components`

ntsekouras replied:
No strong opinions here, although I'm not sure we'd have more `sizes/variants`.

ciampo replied:
A `variant` prop with `"default"` and `"compact"` as allowed values may be more future-proof

jameskoster commented:
<!-- linear:replyTo:2841200244 -->

@jasmussen I think the behavior you describe should be reserved for when the current page is visible in the trail. Otherwise, when collapsing we should show the furthest ancestor and the closest one. I made a [quick prototype](https://codepen.io/jameskoster/full/YPPERep) to describe the behavior.

jameskoster commented:
<!-- linear:isThreadRoot -->

@ntsekouras in the compact variant the ellipsis button increases the height of the overall breadcrumb and the `…` icon looks a bit mis-aligned with the other items. Do we need to use a Button as the trigger? Wondering if we can use a regular text link instead.

jameskoster commented:
<!-- linear:isThreadRoot -->

Not a strong feeling but in most cases I suspect folks will want to visually hide the current page, so that could be the default config.

ntsekouras commented:
>@ntsekouras in the compact variant the ellipsis button increases the height of the overall breadcrumb and the … icon looks a bit mis-aligned with the other items. Do we need to use a Button as the trigger? Wondering if we can use a regular text link instead.

@jameskoster 

>In BreadcrumbsMenu I used DropdownMenu to get the keyboard navigation for free. The issue with that though is that we cannot skip the icon prop, that is used to render a button dialog trigger. We can either add styles there and create a new svg icon, or we could use the Dropdown component and implement the keyboard navigation inside this component

jasmussen commented:
I think the behavior you showed there is the same as I'm proposing:

![Screenshot 2025-04-30 at 15 44 40](https://github.com/user-attachments/assets/c19ba114-c378-4b1f-b381-faf177de8816)

And for me that accounts for both of these configurations:

> Home / ... / Subcategory / Current page

> Home / ... / Subcategory /
> **Current page**

However good to clarify. Does the above make sense?

jameskoster commented:
<!-- linear:replyTo:2842018114 -->

Ah! Should we use [Menu](https://wordpress.github.io/gutenberg/?path=/docs/components-menu--docs) instead? Then I think we can adjust the Trigger.

jameskoster commented:
<!-- linear:isThreadRoot -->

@ntsekouras small detail: is there any benefit to including the separators in the markup, rather than using css pseudo elements?

jameskoster commented:
<!-- linear:isThreadRoot -->

@ntsekouras I noticed we're using `Button` for each breadcrumb item, is that necessary? I think it would be easier to style if we used regular links.

ntsekouras commented:
>In BreadcrumbsMenu I used DropdownMenu to get the keyboard navigation for free. The issue with that though is that we cannot skip the icon prop, that is used to render a button dialog trigger. We can either add styles there and create a new svg icon, or we could use the Dropdown component and implement the keyboard navigation inside this component

>Ah! Should we use [Menu](https://wordpress.github.io/gutenberg/?path=/docs/components-menu--docs) instead? Then I think we can adjust the Trigger.

Menu is a private component. That's another alternative indeed, but I'm not sure if we want `unlock` it and expose it (or not)  through `@automattic/components`? We've used this components in lot's of places in GB and DataViews and it seems to work well IMO. What do you think @ciampo @youknowriad ?

ntsekouras committed:
feedback part 1

youknowriad commented:
Exposing `Menu` component as stable seems like a good thing to me, assuming it's the one we want to encourage folks to use. Would love to also maybe reexport VStack/HStack without prefixes, the Tab component too may be exposed as stable. I'll defer to @ciampo and @mirka though.

ntsekouras committed:
always show parent item

ntsekouras committed:
minor updates

ntsekouras committed:
add some basic unit tests

ntsekouras committed:
add changelog entry

ntsekouras committed:
minor changes

jameskoster commented:
<!-- linear:replyTo:2842387572 -->

From the design side of things `Menu` is definitely the most up-to-date and preferred option.

youknowriad commented:
Looks good to me but I'd appreciate some eyes from our components friends :) cc @ciampo @aduth @mirka

ntsekouras committed:
Unlock `Menu` and use it in `Breadcrumbs`

ciampo commented:
<!-- linear:replyTo:2842387572 -->

Excuses for the delay. I'm catching up with a lot of unread GH / linear notifications, I plan on taking a look either later today, or tomorrow morning at the latest.

jameskoster committed:
Style adjustments

jameskoster commented:
<!-- linear:isThreadRoot -->

I pushed a couple of small style tweaks. As always I'd welcome more feedback but I think this is in a good spot in terms of design. 

Noting that I lightened the separator to match Figma, and it is my understanding this as a decorative element it doesn't need to meet any contrast requirements.

ciampo commented on packages/components/src/menu/index.tsx:
I have some questions about how we're unlocking and re-exporting `Menu`.

In Gutrenberg, we're deliberately using `Object.assign` (see [guidelines](https://github.com/WordPress/gutenberg/blob/8ac0e792402cd2102f2ff80f6a32172a5aff3f37/packages/components/CONTRIBUTING.md#naming-conventions)) to add compound sub-components as properties of the top-level component (if I remember correctly, the main reason was for TypeScript/Storybook compat and general devX)

The way we're re-constructing the `Menu` object doesn't use `Object.assign` and doesn`t add all the JSDocs that are lost because of the lock/unlock.

![Screenshot 2025-05-06 at 14 21 47](https://github.com/user-attachments/assets/a54f5a17-9f61-4cc7-98ef-08eb70a07c1d)

![Screenshot 2025-05-06 at 14 21 55](https://github.com/user-attachments/assets/9bcce12b-a231-4e87-81cb-29776105c0a2)

Since we're importing the types directly from the `@wordpress/components/src/menu/types`, we may as well import the original `Menu` component ?

```tsx
export { Menu } from '@wordpress/components/src/menu';
```

Or otherwise, we could just re-export the "unlocked" `Menu` and add types, without re-assigining the compound components.

Another note about devx: both the original JSDocs and README are missing, meaning that any consumers of `Menu` through `@automattic/components` are not getting docs and would need to manually look for the `@wordpress/components` docs instead.

Finally, what's the plan about iterating on this component? What do we do if we need to make any changes? If the answer is "we make a local copy", then should we just make a copy directly?

cc @Automattic/team-calypso

tyxla replied:
> Finally, what's the plan about iterating on this component? What do we do if we need to make any changes? If the answer is "we make a local copy", then should we just make a copy directly?

For now, that's not an issue. If we need to make changes, we can make a local copy in `@automattic/components` and iterate in it, and make it the canonical version to use across a8c.

I agree with @ciampo on the rest he said 👍

ntsekouras replied:
I had a call with @ciampo about this and the result is this [commit](https://github.com/Automattic/wp-calypso/pull/102998/commits/c6415b40ec7e05d124140a4b68747678a725cc59). 

I'll mark this as resolved because it refers to outdated code, and let's discuss if needed on the new code.

ciampo commented on packages/components/src/breadcrumbs/style.scss:
```diff
@@ -0,0 +1,36 @@
+@import "@wordpress/base-styles/mixins";
+@import "@wordpress/base-styles/variables";
```
This is more of a general consideration (as I see this is widespread across the whole repo), but is using Sass variables and mixins an "official" best practice (that we'd like to encourage), or something that happened organically (and we may want to consider going forward)? Asking mostly @mirka @jameskoster @jasmussen

jameskoster replied:
Given the current direction, utilising core design tokens make sense to me. But I suspect there may be a nuance to the question that I'm not picking up 🙂

ciampo replied:
I guess the considerations are:

- by using those variables, we have an explicit dependency of Gutenberg and the WP design system — which is not bad _per se_, but something we should be aware of;
- in terms of design and DS specs, we will need to tidy variables up a bit between the WP DS, the A8C DS, color studio, and any other spare variables. I assume that the new central place / source of truth will be the DS color tokens, ie. the same provided by the `Theme` package?

jameskoster replied:
I suppose the big unknown is how much the theme package will ultimately be responsible for. Is it just colors, or other aspects of design too? Theoretically it could drive typography, spacing, density, radius, and more…

ciampo commented on packages/components/src/breadcrumbs/index.tsx:
```diff
@@ -0,0 +1,146 @@
+	return (
+		<HStack
+			as="nav"
+			className="breadcrumbs"
```
Let's namespace our classnames — something like `a8c-components-breadcrumbs` maybe? I remember @mirka proposing a standard.

Alternatively, if CSS modules build correctly, we could just use CSS modules instead?

ciampo commented on packages/components/src/breadcrumbs/index.tsx:
```diff
@@ -0,0 +1,146 @@
+function BreadcrumbItem( { item: { label, href, onClick } }: { item: BreadcrumbItemProps } ) {
+	return (
+		<span className="breadcrumb__item-wrapper">
+			<a href={ href } onClick={ onClick } className="breadcrumb__item">
```
Should we render a `span` if the `href` is not defined or empty?

ntsekouras replied:
Actually it makes no sense for a breadcrumb item not to have an `href`. I'll update the prop to be required for now, but if design thinks this should also work for buttons too, we should update the code to render different tags conditionally. --cc @jameskoster @jasmussen

jameskoster replied:
@ntsekouras are you asking whether the breadcrumb should support using Buttons for the trail items as well as links? I can't really think of a need for that. Apologies if I'm mis-interpreting the question.

ntsekouras replied:
That was exactly my question. We can move forward with link then and revisit if we even need buttons in the future.

ciampo replied:
I'm with requiring `href` for now, although I think that we may want to make it optional in the future. For example, it may be useful to render breadcrumbs even if we don't want them to be clickable (ie. simple text), as additional piece of information?

ciampo commented on packages/components/src/breadcrumbs/index.tsx:
```diff
@@ -0,0 +1,146 @@
+					{ items.map( ( item, index ) => (
+						<Menu.Item
+							key={ `${ item.label }-${ index }` }
+							render={
+								<a href={ item.href } onClick={ item.onClick } className="breadcrumb__item" />
+							}
```
We should render an `<a />` tag only if an `href` is defined

```suggestion
							onClick={ item.onClick }
							className="breadcrumb__item"
							render={
								item.href ? ( <a href={ item.href } /> ) : undefined
							}
```

ciampo replied:
Also, is adding the `breadcrumb__item` class necessary? The styles should be already be coming from `Menu`

ntsekouras replied:
Good point about the css class. Regarding the `href` it's similar to my [comment](https://github.com/Automattic/wp-calypso/pull/102998/files#r2075960246). Let's see if we have to conditionally render different tags.

ciampo commented on packages/components/src/breadcrumbs/index.tsx:
```diff
@@ -0,0 +1,146 @@
+
+export const Breadcrumbs = forwardRef( UnforwardedBreadcrumbs );
+
+/**
+ * The breadcrumb component provides a secondary navigation aid that shows users their current
+ * location within a site's or application's hierarchy. It helps users understand the structure
+ * of the site, retrace their steps, and easily navigate to higher-level pages.
+ *
+ * For accessibility, it is important that the current page is included as the final item in the
+ * breadcrumb trail. This ensures screen reader users receive the full navigational context.
+ *
+ */
```
We should add the JSDoc comment to the named export above, so that it gets picked up both when importing the named export, and the default export.

Side question — do we need a default export?

ntsekouras replied:
>Side question — do we need a default export?

Not really, it's what we usually do and personally I don't think it matters much in most cases. I can remove it here.

ciampo commented on packages/components/src/breadcrumbs/index.tsx:
```diff
@@ -0,0 +1,146 @@
+		setContainerWidth( resizeObserverEntries[ 0 ].borderBoxSize[ 0 ].inlineSize );
+	} );
+	const mergedRefs = useMergeRefs( [ ref, containerRef ] );
+	if ( ! items.length || items.length === 1 ) {
+		return null;
+	}
```
Why are we not showing a one-item breadcrumbs? We could show them and give the user freedom to do that, if needed. Of course, compact mode won't be a thing anyway, and we'd need to decide if hiding the `currentItem` should be allowed.

ntsekouras replied:
We could, but for me it made no sense to have `breadcrumbs` with a single item. I might be wrong though.. --cc @jameskoster @jasmussen

jameskoster replied:
I guess it depends on the context of `items`—does that include the current page? If so, and the single item is the current page then I agree there's no need to display the breadcrumb. But if `items` refers to the interactive links in the trail, and there is only one, I think it should be displayed.

ntsekouras replied:
That's the current behavior, yes. If we only pass a single item, it's the `current` item, so it's not displayed.

ciampo commented on packages/components/src/breadcrumbs/index.tsx:
```diff
@@ -0,0 +1,146 @@
+}
+
+function UnforwardedBreadcrumbs(
+	{ items, showCurrentItem = false, isCompact = false }: BreadcrumbProps,
```
Not a blocker, but for better extensibility, we usually accept and forward all HTML props/attributes for the underlying element

```suggestion
	{ items, showCurrentItem = false, isCompact = false, ...restProps }: BreadcrumbProps & React.HTMLAttributes< HTMLNavElement >,
```

ntsekouras replied:
It seems there is no `HTMLNavElement`. From my quick search I didn't find the alternative. I'll look into it tomorrow more, but if you know it, please share :)

ciampo commented on packages/components/src/breadcrumbs/index.tsx:
```diff
@@ -0,0 +1,146 @@
+) {
+	const scrollWidth = useRef( 0 );
+	const [ containerWidth, setContainerWidth ] = useState( 0 );
+	const containerRef = useResizeObserver( ( resizeObserverEntries ) => {
+		const [ entry ] = resizeObserverEntries;
+		const { inlineSize } = entry.borderBoxSize[ 0 ];
+		/**
+		 * We only need to set the `scrollWidth` once, but we also need to ensure
+		 * that it is only set when there was no previous value and it's different
+		 * than the container's `inlineSize`. This is because the `nav` element expands
+		 * to the available space and in a large container where the `nav` doesn't overflow,
+		 * would have the same value for `scrollWidth` and `inlineSize`.
+		 * We cannot set something like `width: fit-content`, as it would never
+		 * expand (hide the dropdown) in future resizes.
+		 */
+		if ( ! scrollWidth?.current && entry.target.scrollWidth !== inlineSize ) {
+			scrollWidth.current = entry.target.scrollWidth;
+		}
+		setContainerWidth( resizeObserverEntries[ 0 ].borderBoxSize[ 0 ].inlineSize );
+	} );
```
> Noting that in storybook where you have the ability to change the items from controls, the responsive behavior is quirky due to the calculations regarding the compact format. You should refresh and then play around with the resizing.
> @ciampo do you know if this makes sense for storybook or a bug might be hidden in the actual code?

I can confirm the bug that you raised, and I believe that it would happen also outside of Storybook — basically, `scrollWidth.current` doesn't get updated accordingly when the items change (ie. different number of items, different labels, etc).

To fix the behaviour, `scrollWidth.current` needs to be updated more frequently — although the challenge with the current setup is that it won't be possible to measure the updated `scrollWidth` while in compact mode.

My suggestion is that we always render an invisible, off-screen "full width" copy, which we use to measure `scrollWidth` and `containerWidth`.

What do you think?

ntsekouras replied:
I wanted to avoid a second container, but it might not be possible. I used your suggestion here: https://github.com/Automattic/wp-calypso/pull/102998/commits/ce10201f14c59d13828bb9c98eaf9ef70fe2be16

ntsekouras committed:
feedback part 1

ntsekouras committed:
rename `size` to `variant`

ntsekouras committed:
use an offscreen full width component for 'compact' calculations

ntsekouras committed:
update the way we export `Menu`

ciampo committed:
Fix missing "s" in breadcrumbs classname

ciampo committed:
Use forwardRef, rename navRef to ref

ciampo committed:
Move props outside of `render` when possible

ciampo committed:
Tweak JSDocs

ciampo committed:
Use `@wordpress/react-i18n` instead of `@wordpress/i18n`

ciampo committed:
Use variant prop instead of extra internal shouldRenderCompact prop

ciampo committed:
Accept HTML attributes

ciampo committed:
Use resize observer instead of useLayoutEffect

ciampo committed:
Use `inert` instead of `aria-hidden`

ntsekouras commented on packages/components/src/types.d.ts:
```diff
@@ -1,6 +1,19 @@
 
 declare const __i18n_text_domain__: string;
+
+declare module 'react' {
```
@ciampo why do we need this? Also I can't build with the recent changes.. 🤔

ciampo replied:
Still working on a few changes, that's why I haven't left a comment yet.

Do you get build errors locally?

ntsekouras replied:
Yes.

>packages/components/src/horizontal-bar-list/stats-card.tsx(131,5): error TS2322: Type '{ children: (false | Element | undefined)[]; 'aria-hidden'?: true | undefined; inert?: "" | undefined; className: string; }' is not assignable to type 'DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>'.
  Type '{ children: (false | Element | undefined)[]; 'aria-hidden'?: true | undefined; inert?: "" | undefined; className: string; }' is not assignable to type 'HTMLAttributes<HTMLDivElement>'.
    Types of property 'inert' are incompatible.
      Type '"" | undefined' is not assignable to type 'boolean | "true" | "false" | undefined'.
        Type '""' is not assignable to type 'boolean | "true" | "false" | undefined'.


The error seems weird, but it might be related with the `inert` change here.

ciampo replied:
Got it. They are referred to the `inert` types that I added — I already removed them locally and will push a fix soon

ciampo committed:
Fix inert types

ciampo committed:
Move React.HTMLAttributes directly to BreadcrumbProps for better Storybook compat

ciampo committed:
Move aria-label destruct + default value to root component

ciampo committed:
Assistive technlogy fixes and improvements:

- do not mark the current item with aria-hidden
- DOM markup to nav > ul > li > span / a / button

ciampo committed:
Storybook type tweaks

ciampo committed:
Rewrite tests to adapt to new markup and to use more semantic queries

ciampo committed:
Remove custom type by getting around TS error

ciampo committed:
Storybook allow default items controls

ciampo commented:
Hey @ntsekouras 👋  as agreed, I spent some time on this PR and applied a few tweaks. You can review my changes commit-by-commit, but at a high level:

- refactored from using `useLayoutEffect` to using `useResizeObserver` on the offscreen Nav as well — this is because we really need to react to any changes to the width of the nav, and not just the items prop changing. Even a CSS change (such as changing padding, font size, etc) should trigger the calculations. 
- using `inert` alongside `aria-hidden`. Compared to `aria-hidden`, `inert` also makes the HTML subtree non-interactive (useful to make sure that the user cannot focus the `<a />` elements in the offscreen nav)
- changed the DOM structure to `nav > ul > li > a / button / ...`. That's because it's a best practice to use an actual list, rather than having the links / button as normal children or `nav`;
- removed `aria-hidden` from the current item because that would cause assistive technology not to pick it up;
- Updated all classnames to have `a8c-components-breadcrumbs` prefix for better consistency;
- used `forwardRef` for the `BreadcrumbsNav` component;
- moved from `@wordpress/i18n` to `@wordpress/react-i18n` (more on this below);
- changed the props to accept also all HTML attributes;
- rewrote unit test to account for the above changes (and use semantic queries as much as possible).

Let me know if the component looks and works as you expected!

ntsekouras commented:
Thank you so much @ciampo! Code changes look good to me! Now I guess we need a ✅ from someone else :).

tyxla commented on packages/components/src/breadcrumbs/index.tsx:
```diff
@@ -0,0 +1,185 @@
+			<Menu placement="bottom-start">
+				<Menu.TriggerButton
+					className="a8c-components-breadcrumbs__item"
+					render={ <Button size="compact" text="…" label={ __( 'More breadcrumb items' ) } /> }
```
I wonder if the ellipsis should be localized.

tyxla commented on packages/components/src/breadcrumbs/index.stories.tsx:
```diff
@@ -0,0 +1,51 @@
+export const Default: Story = {
+	args: {
+		items: [
+			{ label: 'Home', href: '#' },
```
These `#` don't seem to be working nicely, they're getting the user out of Storybook into an iframe mode.

Should we use `javascript:void(0)` instead?

tyxla reviewed (commented):
This mostly looks good 👍 

One thing: I found it unexpected and inconsistent that we display a trailing breadcrumb item separator sometimes, but not always:

![Screenshot 2025-05-08 at 11 13 19](https://github.com/user-attachments/assets/73212b71-e4e8-4789-bfd1-b9841332d6eb)
![Screenshot 2025-05-08 at 11 12 37](https://github.com/user-attachments/assets/a08339a6-97f4-48d4-aaa6-737f4151057c)
![Screenshot 2025-05-08 at 11 12 33](https://github.com/user-attachments/assets/113f3eff-d4c2-4412-9e7c-095c76bca7a9)

I expected that if we don't display the current item, we also don't display the corresponding separator. 

The rest is minor, see inline comments. I think it should be good to go once we address the trailing item separator inconsistency.

ciampo committed:
Prevent links from leaving storybook

ciampo committed:
localize ellipsis

ntsekouras commented:
>I expected that if we don't display the current item, we also don't display the corresponding separator.

The most common use case would be to have the title right below the Breadcrumbs, so personally I think it's good to have it.  In general for me it's a sign that indicates whether the last breadcrumb item is actually the last item of the path.

--cc @jameskoster @jasmussen

ciampo commented:
@tyxla Addressed the two code comments.

> I expected that if we don't display the current item, we also don't display the corresponding separator.

I guess it's a design choice — a trailing "/" indicated that the current page is hidden in the breadcrumbs, while the lack of indicates that the last item in the breadcrumbs is the current page. Although I understand it's a subtle detail to understand and it may create confusion  (cc @jasmussen @jameskoster )

In any case, I believe that we can make tweaks to the "/" behaviour in a follow up if required.

Other considerations that came to my mind:

- I thought about proposing a compound components approach for this component too, but it would be probably over-engineered, especially if we don't forsee needs for customizing the single breadcrumb items;
- do we foresee any needs for integrating the links in the component with the routing system? I guess, at most we could add a `renderLink` prop which would allow folks to render a different component than the vanilla `<a />` ? (cc @youknowriad @mirka )

ntsekouras commented:
>do we foresee any needs for integrating the links in the component with the routing system? I guess, at most we could add a renderLink prop which would allow folks to render a different component than the vanilla <a /> ? (cc @youknowriad @mirka )

We'll most certainly do, but why would we need a `renderLink`? The consumer should have to construct the breadcrumbs based on the current path and just pass them along.

ciampo commented:
> > do we foresee any needs for integrating the links in the component with the routing system? I guess, at most we could add a renderLink prop which would allow folks to render a different component than the vanilla  ? (cc @youknowriad @mirka )
> 
> We'll most certainly do, but why would we need a `renderLink`? The consumer should have to construct the breadcrumbs based on the current path and just pass them along.

I don't know exactly how the routing system will work, and if using a vanilla `<a />` is enough, or if you'd need a particular `Link` component that is tied with the routing system

jasmussen commented:
> The most common use case would be to have the title right below the Breadcrumbs, so personally I think it's good to have it. In general for me it's a sign that indicates whether the last breadcrumb item is actually the last item of the path.

I would expect these are the two behaviors for breadcrumbs:

<img width="624" alt="Frame 11" src="https://github.com/user-attachments/assets/949722e0-a24c-4859-bde5-8602efecd615" />

One omits the last crumb becuase there's a title right next to it. The other includes the last crumb because there isn't a title next to it. In both cases, the last slash is present.

Deos that answer the question?

tyxla commented on packages/components/src/breadcrumbs/index.stories.tsx:
```diff
@@ -0,0 +1,51 @@
+	args: {
+		...Default.args,
+		items: [
+			{ label: 'Home', href: '#' },
```
Let's fix all the links like we did above

tyxla reviewed (approved):
This looks good from my perspective 👍 

Nice team work, folks! 🥇 

I think we can follow-up with any subsequent improvements in separate PRs - no need to block this one.

ciampo committed:
Swap more links in Storybook

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/17454222

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Hi @ntsekouras, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include this string: `More breadcrumb items`

Thank you in advance!
